src/fsharp/TypeChecker.fs

```diff
@@ -11006,10 +11013,18 @@ and TcAndBuildFixedExpr cenv env (overallPatTy, fixedExpr, overallExprTy, mBindi
     | ty when isStringTy cenv.g ty -> 
         let charPtrTy = mkNativePtrTy cenv.g cenv.g.char_ty
         UnifyTypes cenv env mBinding charPtrTy overallPatTy
+        
+        // When targeting .NET Core or higher:
+        //
+        //    let ptr: nativeptr<char> = 
+        //        let pinned s = str
+        //        (nativeptr)s + GetPinnableReference()
```

Comment:
Just to clarify, the compiler shouldn't perform any arithmetic here. The result of `GetPinnableReference` is a _readonly char\&_, and that _char\&_ itself should be pinned. The string instance itself should not be pinned.

Reply:
Hmmm. @dsyme wrote this originally, I assume he had a good reason for doing this, but yes it shouldn't need to do that.

Reply:
Clarification to my comments: the `GetOffsetToStringData` pattern works by having the caller pin the string instance, then compute `addrof(string) + GetOffsetToStringData`, then expose the resulting sum as a _char\*_. So the addition is expected there.

The `GetPinnableReference` pattern works by providing the _char\&_ directly, which the caller then pins.

Reply:
Gotcha, I'll take another look at this